コメント機能課題提出

keiz1213 opened the pull request:
よろしくお願いいたします。

keiz1213 committed:
モデルとコントローラー作成

1.reportモデル作成
2.reportsコントローラー作成
3.UserとReportの関連付け

keiz1213 committed:
日報投稿機能追加

keiz1213 committed:
view一部修正

1.他ユーザーの日報一覧画面で「編集」、「削除」、「新規作成」が表示されないようにした

keiz1213 committed:
コメント投稿機能追加

1. コメントモデル作成
2. ポリモーフィック関連付け(ユーザー、本、日報)
3. コメント入力フォーム作成
4. コメント一覧作成
5. 不要なファイル削除

keiz1213 committed:
HTML5の必須チェック機能漏れを修正

1.app/views/comments/_form.html.erb

keiz1213 committed:
翻訳修正

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def create
+    @comment = @commentable.comments.build(comment_params)
+    @comment.save
```
HTML側で `required: true` しているのでバリデーションエラーが発生しない、という前提とするなら `@comment.save!` とした方が良いですね。
最悪、何らかの原因で保存に失敗した場合に例外が発生して気づくことができるためです 👍

yuuu replied:
[こちら](https://github.com/keiz1213/fjord-books_app/pull/6/files#diff-34e52388abd1ae0658edd56e12cd6d935fac98673361084ba07a99ca4bb836e7R15) も `@comment.update!` を使うと良いですね。

keiz1213 replied:
なるほどです。万が一失敗したときのことを考慮していませんでした😅修正しておきます。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def update
+    @comment.update(comment_params)
+    redirect_to @commentable_type, notice: t('controllers.common.notice_update', name: Comment.model_name.human)
```
`@commentable_type` ではなく `@commentable` だと何か不都合がありましたでしょうか？

keiz1213 replied:
コメントを更新して該当の`book`または`report`にリダイレクトする際に、「リダイレクト先が無い」というエラーが出るため`@commentable_type`にしています。

#### リダイレクト先を`@commentable`にしてコメントを更新した時に出るエラー
![Action_Controller__Exception_caught](https://user-images.githubusercontent.com/72614612/144697845-9fd9c18f-f5a5-452e-a411-6c9401d7d92e.png)

yuuu replied:
@keiz1213 
こちら見逃していました。フォローいただきありがとうございます。
`@commentable` が `nil` となっているようですが、`set_commentable` にてセットされているかどうか確認されていますでしょうか？

keiz1213 replied:
初歩的なこと見逃していました・・😅
確かに、`update`アクションでは`set_commentable`で`@commentable`がセットされていませんでした。
`set_commentable`は新規コメント投稿時に`Book`または`Report`に応じてコントローラーを分けてセットされるように作っています。
現時点で、更新するためのURLは`/comments/comment_id`にしているので、URL`/reports/report_id/comments/comment_id`のようにすれば`update`アクションでも`@commentable`がセットされ、こちらにリダイレクトできるようになると思うのですがこのように修正したほうがよろしいでしょうか？

下記のようにとりあえず修正してみましたが正常に動きました。

```erb
# app/views/comments/_form.html.erb

<%= form_with model: [@commentable_type, @comment] do |f| %>  
  <%= f.text_area :content, required: true %>  
  <div class="actions">  
    <%= f.submit %>  
  </div>  
<% end %>
```

```ruby
# app/controllers/comments_controller

  def update
    @comment.update!(comment_params)
    redirect_to @commentable, notice: t('controllers.common.notice_update', name: Comment.model_name.human)
  end
```

yuuu replied:
なるほど、erb側でのformのリンクが誤っていたということですね。せっかく作ったコントローラを活用できていないのはもったいないので、上記の通り修正をお願いします🙏

keiz1213 replied:
ありがとうございます！修正しておきます。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def show
+    @user = User.find(params[:id])
```
Reportsのshowですが `User.find(params[:id])` で大丈夫でしょうか？
もし、Reportsに紐付いているユーザーを取りたいなら `@report.user` で取れると思います。

keiz1213 replied:
確かに`@report.user`でユーザーが取れました。一応、`User.find(params[:id])`でも同じように動作するのですが、`@report.user`の方が処理速度が早い(?)という点でこちらの方が良いのでしょうか？

yuuu replied:
@keiz1213 
なるほど。同じように動作する、とのことですが `params[:id]` に入っているIDはReportのIDではないでしょうか？
ReportのIDとUserのIDがたまたま一致していればそれらしく動作するとは思いますが、そうでない場合は間違ったUserを取得するか、Userを取得できずに404エラーになると思われます。

keiz1213 replied:
確かにです！！！
`binding.irb`で確認してみたところ、`User.find(params[:id])`に入っている`id`はReportの`id`が入っていました！😅

このままで同じように動いていた理由は、そもそもこの`@user`をビューで使用していませんでした・・・🙏
本来この`@user`はレポートの詳細ビューにてそのレポートの作者のみ編集ができるように、`current_user`との比較用に用意していました。ですが何故か自分は実際のビューで`@report.user`でユーザーを取ってきてそれを使用していました。なので`User.find(params[:id])`で取得した`@user`が問題にならなかったようです・・🙏
```ruby
<% if current_user == @report.user %>
  <%= link_to t('views.common.edit'), edit_report_path(@report) %> |
<% end %>
```

yuuu commented on app/controllers/users/reports_controller.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
```
このファイルは必要でしょうか？使いみちがよくわからなかったので念の為確認です。

keiz1213 replied:
「全レポート一覧」と「個別ユーザーの日報一覧」を分けて表示しようと考えました。そのため、全レポートの一覧を`/reports`というURLに対して`reports_controller`で表示し、個別ユーザーの日報一覧を`/users/user_id/reports`に対してこの`users/reports_controller`で表示しようとして作成しました。

yuuu replied:
こちらのファイルの必要性について確認をお願いします🙏

yuuu replied:
なるほど、確かに `routes.rb` にも定義が追加されていましたね。問題ないと思います 👍

yuuu commented on app/models/report.rb:
```diff
@@ -0,0 +1,8 @@
+class Report < ApplicationRecord
+  has_many :comments, as: :commentable, dependent: :destroy
+  belongs_to :user
+  default_scope -> { order(created_at: :desc) }
```
`default_scope` は一見便利そうに見えるのですが、Controller側の処理を追ってもORDERしていることに気づきにくくなるので、諸刃の剣ですね。
今回はController側で明示的にORDERしていたので指定は不要だと思いますがいかがでしょうか？

keiz1213 replied:
なるほどです。確かにControllerにて明示的に指定していたのでこちらの`default_scope`は不要ですね。気づきにくくなるというのは確かにそうだなと思いました💡一点質問なのですが、降順昇順の指定をつけたい時にModel側とController側どちらで指定する方がベターという基準はありますでしょうか？

yuuu replied:
基本的にはControllerにて `Report.order(created_at: :desc)` のようにコントローラ側で指定します。ただ、このメソッドチェーンが長くなる場合、もしくは予め長くなった場合を見越して、Model側に別途scope(default_scopeではなくscope)を定義することが多いですね。

scopeを定義してもConrtoller側でそのscopeを呼び出す必要があるので、指定をする、という意味ではController側ということになります。

keiz1213 replied:
回答ありがとうございます！
基本的にはController側で必要に応じて指定する。もし条件が多くなって長くなりそうだったらモデル側で`scope`で定義してそれを呼び出す感じですね💡
`default_scope`と`scope`ついて改めて調べてみましたが、そもそも基本的な理解ができていなかったみたいです。
確かに、`default_scope`で指定していると、「利便性は良いけど気づきにくくなってメンテナンス性が落ちる」ということがしっかり理解できました💡

yuuu commented on app/views/books/show.html.erb:
```diff
@@ -22,3 +22,12 @@
+<%= form_with model: [@book, Comment.new] do |f| %>  
+  <%= f.text_area :content, required: true %>  
+  <div class="actions">  
+    <%= f.submit "コメントをする", class: 'btn btn-warning px-5' %>  
```
I18n対応が漏れているようなので対応をお願いします🙏

yuuu replied:
以下箇所も同様ですね。

- https://github.com/keiz1213/fjord-books_app/pull/6/files#diff-d1362e358cc8657c022cfc98851a57ee03e3790feb20ba00b8e3da772950286eR41
- https://github.com/keiz1213/fjord-books_app/pull/6/files#diff-9d88a8a75acf290954deff29565d6b3a7a2f120f439b5ea416b6403eed88e993R4
- https://github.com/keiz1213/fjord-books_app/pull/6/files#diff-ddbaaabeabe0d5e4f1950f228bf17d58866879a35dd04471d272174a1143186fR23

keiz1213 replied:
詰めが甘かったです・・・🙏こちら修正しておきます。

keiz1213 committed:
指摘事項修正

1. default_scope削除
2. Reports#showのユーザインスタンス修正
3. コメントの作成、保存が失敗した際に例外を出すようにした
4. コメントに関するi18n対応

keiz1213 committed:
不要なファイル削除

1. 謎のテスト用ファイルが作成されていたので削除した

yuuu commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,27 @@
+<%= form_with(model: report, local: true) do |form| %>
+  <% if report.errors.any? %>
+    <div id="error_explanation">
+      <h2><%= pluralize(report.errors.count, "error") %> prohibited this report from being saved:</h2>
```
この部分のi18nが漏れているようですね

keiz1213 replied:
こちら修正しました🙏

yuuu commented on config/initializers/time_formats.rb:
```diff
@@ -0,0 +1 @@
+Time::DATE_FORMATS[:datetime_jp] = '%Y年 %m月 %d日 %H時 %M分'
```
日時のローカライズは `I18n.l` を使ったほうが良いですね。
https://railsguides.jp/i18n.html#%E6%97%A5%E4%BB%98%E3%83%BB%E6%99%82%E5%88%BB%E3%83%95%E3%82%A9%E3%83%BC%E3%83%9E%E3%83%83%E3%83%88%E3%82%92%E8%BF%BD%E5%8A%A0%E3%81%99%E3%82%8B

keiz1213 replied:
こんな便利で簡単な方法があったとは・・・修正しました！

keiz1213 committed:
指摘事項修正(2)

1. 日報登録に関するエラーメッセージのI18n対応
2. 日時のローカライズにI18n.lを使用
3. config/initializers/time_formats.rb を削除

keiz1213 committed:
form_withのリクエストURLを変更

1. リダイレクト先を@commentableにするため
